fact_extractor/plugins/unpacking/draytek_vigor_167/code/draytek_vigor_167.py

```diff
@@ -0,0 +1,35 @@
+'''
+This plugin extracts the squashfs root filesystem in Draytek Vigor 167 firmware containers.
+'''
+import struct
+from pathlib import Path
+from common_helper_files import write_binary_to_file
+
+NAME = 'Draytek Vigor 167'
+MIME_PATTERNS = ['firmware/draytek-vigor-167']
+VERSION = '0.1'
+
+header_layout = '> 4s I I I 64x I I 49x I I 117x'
+header = struct.Struct(header_layout)
+
+
+def unpack_function(file_path, tmp_dir):
+    '''
+    file_path specifies the input file.
+    tmp_dir must be used to store the extracted files.
+    Optional: Return a dict with meta information
+    '''
+    with open(file_path, 'rb') as f:
+        signature = header.unpack(f.read(header.size))
+        f.seek(signature[1] + signature[4])
+        squashfs = f.read(signature[5])
```

Comment:
There are a few things that could go wrong here:

1. [unpack](https://docs.python.org/3/library/struct.html#struct.Struct.unpack) requires the input buffer size to be in line with the format's byte requirements. If `f.read(header.size)` returns some `n < header.size` (which might happen if the file is smaller than the header), this will raise an exception.
2. `read` and `seek` calls might raise `IOError`

Reply:
1. I agree. I think there are two ways to fix this.
1.1 Catch the exception as a `struct.error` and abort unpacking.
1.2 Update the Draytek Vigor 167 mime type in [fact_helper_file/mime/custom_mime_firmware_containers](https://github.com/fkie-cad/fact_helper_file/blob/c4c6341d9cfc83e9b4e61e887ec3c2cb98aa8dc4/fact_helper_file/mime/custom_mime_firmware_containers#L718) to include the padding. If the file is too small aka the header does not match perfectly, the extractor wouldn't use this unpacking plugin, right?

2. True. I will put this `with` section inside a try-except block.

Reply:
We agreed on the first solution (1.1) - changed in 18def7084d7ff5a39a7f4cb71a9f68c599bb7a0c